src/libraries/Microsoft.PowerFx.Core/Types/DType.cs

```diff
@@ -3572,5 +3572,68 @@ public bool TryGetSimilarName(DName name, FieldNameKind fieldNameKind, out strin
             return similar != null &&
                    comparer.Distance(similar) < (name.Value.Length / 3) + 3;
         }
+
+        /// <summary>
+        /// Try to union all table child types and checks if any coercion is necessary. Meant to be called from within table type check loop.
+        /// </summary>
+        /// <param name="argType">Current table child type.</param>
+        /// <param name="arg">Child node.</param>
+        /// <param name="errors">Error container from caller.</param>
+        /// <param name="returnType">Composed new type.</param>
+        /// <param name="needCoercion">Checks if child node needs to be coerced.</param>
+        /// <returns></returns>
+        public bool TryUnionWithCoerce(DType argType, TexlNode arg, IErrorContainer errors, out DType returnType, out bool needCoercion)
+        {
+            var isValid = true;
+
+            KeyValuePair<string, DType> schemaDifference = new KeyValuePair<string, DType>(null, Invalid);
+            DType schemaDifferenceType = null;
+            DType coercionType = null;
+
+            returnType = null;
+            needCoercion = false;
+
+            // Deferred and void types are not allowed in tables.
+            var isChildTypeAllowedInTable = !argType.IsDeferred && !argType.IsVoid;
+
+            if (!isChildTypeAllowedInTable)
+            {
+                errors.EnsureError(DocumentErrorSeverity.Severe, arg, TexlStrings.ErrTableDoesNotAcceptThisType);
+                return false;
+            }
+
+            if (!IsValid || (IsRecord && Equals(EmptyRecord)))
+            {
+                returnType = argType;
+            }
+            else if (CanUnionWith(argType))
+            {
+                returnType = Union(this, argType);
+            }
```

Comment:
I understand that this is existing pattern but this essentially calling Union operation twice for success case.

Instead it could call Union(ref fError, ...) directly.